adrs/0004-accessibility-testing-strategy.md

```diff
@@ -0,0 +1,40 @@
+# Stacks Accessibility Testing Strategy
+
+**Date:** 2022-12-21
+
+**Related PR:** [#1217](https://github.com/StackExchange/Stacks/pull/1217)
+
+## Issue
+We currently rely on manual testing, third-party audits, and user reports to identify and fix accessibility issues. The only semi-automated testing we currently Furthermore, while we've generally aspired to WCAG 2.1 AA compliance, we haven't codified the accessibility success criteria for our components.
+
+**A defined success criteria and automated accessibility testing will reduce the effort needed to build and maintain accessibile components.**
+
+## Decision
+
+Introduce the following success criteria for the accessibility Stacks components:
+
+- All components should meet or exceed [WCAG 2.1](https://www.w3.org/TR/WCAG21/) level AA success criteria
+- Components rendered within high contrast mode should meet or exceed [WCAG 2.1 AAA success criteria for contrast](https://www.w3.org/WAI/WCAG21/Understanding/contrast-enhanced.html)
+
+Introduce the following automated accessibility testing for Stacks:
+
+<!-- TODO: Choose *tbd* accessibility testing library to test components in isolation ([wtr](https://modern-web.dev/docs/test-runner/commands/#accessibility-snapshot) is my current top choice, but I'm also exploring other options mentioned in [this PR comment thread](https://github.com/StackExchange/Stacks/pull/1195#discussion_r1037616024)) -->
```

Comment:
Good point. Since we mainly provide CSS classes atm, it's a little tricky to idealize how we can best test for accessibility. We can't enforce how a consumer structures HTML, so we really need to a) structure our component (and their styles) to encourage semantic markup, b) provide examples and guidance in the docs, and c) ensure our component styles (and interactivity, where applicable) meets our accessibility expectations. Not all of this is reasonably testable though.

For our components, I think we could really only test for legibility (min contrast, min font size). As for the documentation, I can think of a couple non-mutually-exclusive options on how to test to setup consumers for success:

- Integrate existing Mabl tests in CI to test our docs and report failures above a certain threshold
- Add tests for each example of a component to ensure our examples meet a11y expectations

---

### Testing examples

> We need an engine that checks for rules violation in the page like axe-core (for example via this chai extension https://open-wc.org/docs/testing/chai-a11y-axe/).

I think this would work for testing the examples we include in our docs.

### Testing components

Legibility seems a little tricky to test dependably. AFAICT, the accessibility testing included in `@open-wc/testing` doesn't seem to accurately test for contrast (though I'd love for this to be proven wrong. I've yet to find an easy drop-in solution for contrast checking, but we could grab computed values in the visual tests and compare them with something like https://github.com/bbc/color-contrast-checker. Basically, something like:

```js

var ccc = new ColorContrastChecker();

const backgroundColor = window.getComputedStyle(button).backgroundColor;
const color = window.getComputedStyle(button).color;
const fontSize = window.getComputedStyle(button).fontSize;

expect(ccc.isLevelAA(backgroundColor, color, fontSize)).to.be.true;
```

Feels kinda cumbersome and potentially frail, so I'm not sold on that approach.

Reply:
The chai-a11y-axe plugin used by open-wc is running axe-core under the hood. It looks like axe-core scans should report a11y violations for color contrasts but I agree with you that we need to investigate further to be 100% sure. 
https://github.com/dequelabs/axe-core/blob/develop/lib/rules/color-contrast.json

We could also manually run the axe core scan. Perhaps we would have a better degree of configurability
https://github.com/dequelabs/axe-core#getting-started